testsuite/tests/compiler-libs/test_longident.ml

```diff
@@ -0,0 +1,104 @@
+(* TEST
+   flags = "-I ${ocamlsrcdir}/parsing"
+   include ocamlcommon
+   * expect
+*)
+
+module L = Longident
+[%%expect {|
+module L = Longident
+|}]
+
+let flatten_ident = L.flatten (L.Lident "foo")
+[%%expect {|
+val flatten_ident : string list = ["foo"]
+|}]
+let flatten_dot = L.flatten (L.Ldot (L.Lident "M", "foo"))
+[%%expect {|
+val flatten_dot : string list = ["M"; "foo"]
+|}]
+let flatten_apply = L.flatten (L.Lapply (L.Lident "F", L.Lident "X"))
+[%%expect {|
+>> Fatal error: Longident.flat
+Exception: Misc.Fatal_error.
+|}]
+
+let unflatten_empty = L.unflatten []
+[%%expect {|
+val unflatten_empty : L.t option = None
+|}]
+let unflatten_sing = L.unflatten ["foo"]
+[%%expect {|
+val unflatten_sing : L.t option = Some (L.Lident "foo")
+|}]
+let unflatten_dot = L.unflatten ["M"; "N"; "foo"]
+[%%expect {|
+val unflatten_dot : L.t option =
+  Some (L.Ldot (L.Ldot (L.Lident "M", "N"), "foo"))
+|}]
+
+let last_ident = L.last (L.Lident "foo")
+[%%expect {|
+val last_ident : string = "foo"
+|}]
+let last_dot = L.last (L.Ldot (L.Lident "M", "foo"))
+[%%expect {|
+val last_dot : string = "foo"
+|}]
+let last_apply = L.last (L.Lapply (L.Lident "F", L.Lident "X"))
+[%%expect {|
+>> Fatal error: Longident.last
+Exception: Misc.Fatal_error.
+|}]
+let last_dot_apply = L.Ldot (L.Lapply (L.Lident "F", L.Lident "X"), "foo")
+[%%expect {|
+val last_dot_apply : L.t =
+  L.Ldot (L.Lapply (L.Lident "F", L.Lident "X"), "foo")
+|}];;
+
+let parse_empty = L.parse ""
+[%%expect {|
+val parse_empty : L.t = L.Lident ""
+|}]
+let parse_ident = L.parse "foo"
+[%%expect {|
+val parse_ident : L.t = L.Lident "foo"
+|}]
+let parse_dot = L.parse "M.foo"
+[%%expect {|
+val parse_dot : L.t = L.Ldot (L.Lident "M", "foo")
+|}]
+let parse_path = L.parse "M.N.foo"
+[%%expect {|
+val parse_path : L.t = L.Ldot (L.Ldot (L.Lident "M", "N"), "foo")
+|}]
+let parse_complex = L.parse "M.F(M.N).N.foo"
+[%%expect {|
+val parse_complex : L.t =
+  L.Ldot (L.Ldot (L.Ldot (L.Ldot (L.Lident "M", "F(M"), "N)"), "N"), "foo")
```

Comment:
`"F(M"), "N)"` really? Yirk.

Reply:
I missed that, but yes `Longident.parse` only parses the subset of `Longident` without application.

Reply:
Can you (or @gasche) comment on what you think should be done then?
Should we change `Longident.parse`? Should we remove the test? Add a comment on the test?

Note that a few lines below we're converting back that broken `Longident` to a string, and everything looks "fine". I don't think we should do that.

Reply:
Personally I think that `Longident.parse` should be fixed to be complete. In fact I hadn't really looked at the test output (otherwise I would have said something here); this continuing experience of surprises makes me half-glad half-sorry that I insisted on writing unit tests for compiler-libs feature.

Reply:
Going down this rabit's hole, making `Longident.parse` complete also means udapting the call in `makedepend.ml` and `typemod.ml` to reject applications in `-open` arguments .

Reply:
I tried extending Longident.parse but it's tedious and I have to decide what to do on parse failure, so I gave up for this PR. I will add a comment to the test to note that this is a known bug, and move on. @trefis do you agree that the PR can go in with this compromise?

Reply:
> @trefis do you agree that the PR can go in with this compromise?

Since you also stopped converting the result of that call to a string I'm fine with for this PR to go in, yes.

Reply:
An intermediate fix would be to make `Longident.parse` fail upon seeing a parenthesis.

Reply:
I think an option returning version might be better since in the two current use cases of `Longident` the error handling should be specialized:
- In `Typemod`, the compiler should warns that `-open F(X)` is not valid (should the input probably be validated earlier?)
- in `toplevel/genprintval.ml`, the extension printer should abort the search for extensions constructor to handle cases like
```OCaml
module F(X:sig end) = struct type t = .. type t+= A end
module E = struct end
let x : F(E).t = let module M = F(E) in M.A;;
```
> val x: <extension>
( currently, the printer searches for a module `Lident F(E)` and fails)

Reply:
> In Typemod, the compiler should warns that -open F(X) is not valid 

That won't stay true for very long: I'm about to send a refreshed version of the "extended open" PR which allows this.

Reply:
> I'm about to send a refreshed version of the "extended open" PR which allows this.

I think that's only true for `.mli` files, for `.ml` files you probably still need a warning.

Reply:
In this case, would it not be simpler to expose the corresponding parser rule (`module_expr`?) from the parser and use `constr_longident` for the extension printer?